### ISSUE TITLE
[windows] lots of tests running in parallel may cause "leaky" warnings

### ISSUE THREAD

**Issue by zaneduffield** [nextest-rs/nextest#1469]
When I run my tests with `datatest-stable` (0.2.9), using `nextest` as the runner, a semi-random number of tests are marked as leaky.

This seems to only depend on how many tests the crate has, and not what the tests are doing.

I have an example repo here: https://github.com/zaneduffield/datatest_leak_issue
That repo has 1000 test files, which are used with a no-op test function.
When I run `cargo nextest run` on it, about 40% of the tests are marked as 'leaky'.
One issue that _may_ be related is how slow this is: ~17s on my machine compared to ~0.2s with `cargo test`.

**Comment by asomers** [comment]
I don't think this is a Windows problem.  I'm seeing it when I run Mockall's test suite (https://github.com/asomers/mockall) with cargo-nextest on FreeBSD 15.0-CURRENT amd64, on a system with 16 hyperthreading cores.  Notably, mockall's test suite does not contain _any_ tests that spawn processes.  I should mention that Mockall's test suite contains 171 binaries.

But if I add `-j1` to the `cargo nextest run` command line, then there are no failures.  In fact, the failure probability is closely related to the value of the `-j` option.  Roughly:
| -j value  | time to test failure |
| ----------- | --------------------------- |
| 4            | never?                       |
| 8            | a few minutes          |
| 16          | a few dozen runs    |
| 32          | 3-5 runs                    |
| 171        | instant                      |

**Comment by sunshowers** [comment]
@asomers interesting. I was wondering whether it was related to not being able to set `CLOEXEC` atomically on pipes, but that's supported on FreeBSD: https://github.com/rust-lang/rust/blob/779e19d8baa3e3625bd4fc5c85cbb2ad47b43155/library/std/src/sys/pal/unix/pipe.rs#L22-L33

**Comment by asomers** [comment]
Correct @sunshowers .  FreeBSD has supported O_CLOEXEC ever since pipe2 was first added to FreeBSD 10.0, which predates Rust itself.

**Comment by sunshowers** [comment]
Thanks. The reason I suspected CLOEXEC was that we used to see leaks on illumos, but those leaks were completely addressed once we added CLOEXEC support to the Rust standard library on illumos (which makes sense, the issue was pipes being leaked in that little window between creation and CLOEXEC being set).

The leaky message shows up if either the stdout pipe or the stderr pipe sent to the tests aren't marked closed within 100ms. I think if you can repro it, some `truss` output that follows all the child processes would be helpful to see what's going on.

We may just have to bump the leak timeout on some platforms like Windows and FreeBSD.

**Comment by asomers** [comment]
I can easily reproduce the bug.  However, I *cannot* reproduce the bug when running with either `truss` or `ktrace`.

**Comment by sunshowers** [comment]
Could you try bumping down the leak timeout to say 50ms?

Also try a few other timeouts with and without truss, e.g. 200ms, 400ms and so on.

https://nexte.st/docs/features/leaky-tests/

**Comment by asomers** [comment]
I don't think that changing the timeout is going to fix this bug.  At best, it will make it less likely to happen.

**Comment by sunshowers** [comment]
Yes, I'm aware, but it would be helpful to get a sense of what parameters are more likely to cause this to happen. Not reproducing under truss or ktrace is strange. Could be a nextest bug, a Tokio bug, a libc bug, a kernel bug, or just a mismatch in expectations. (All of these have been encountered in the past.)

**Comment by asomers** [comment]
Well, just now I _was_ able to reproduce the bug with truss, by setting the timeout down to 5ms.  But the truss file is 20 MB, and I don't know what to look for.  Any pointers?

**Comment by sunshowers** [comment]
Thanks. For the test that was leaky, it would be good to look for events that happen around the process exiting. The rule for leaky tests is:

* wait for the process to exit (done by Tokio via receiving SIGCHLD and running through all spawned children to see which ones exited)
* then, wait 100ms (or the leak timeout duration) for the stdout and stderr file descriptors to close
* mark the test as leaky if they don't close within that duration

In normal use, the stdout and stderr fds should always be closed before the child process exits in a way detectable by the parent. That seems to be how it is on Linux, macOS and illumos. If that's not happening on FreeBSD, it's pretty strange.

I've been thinking about adding some DTrace probes to nextest via https://github.com/oxidecomputer/usdt (which apparently works on FreeBSD). You're welcome to add some to try and get a more precise ordering.

**Comment by sunshowers** [comment]
I've added some basic USDT probes on main. If you install nextest from main:

```
cargo install --git https://github.com/nextest-rs/nextest cargo-nextest --locked
```

Then trace with (command might be slightly different):

```
sudo dtrace -x strsize=512 -Z -n '*:cargo-nextest::test-attempt-done { printf("%s\n", copyinstr(arg0)); }'
```

You should see JSON-encoded output:

```
 16  75522 _ZN14nextest_runner6runner8executor15ExecutorContext8run_test28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h545:test-attempt-done {"ok":{"binary_id":"clap::macros","test_name":"arg_impl::string_literal","attempt":1,"total_attempts":1,"result":"pass","exit_code":0,"duration_secs":0.00864511,"leaked":false,"time_to_close_fds_secs":1e-7,"stress_current":null,"stress_total":null}}
```

For fds that close within the timeout, the `time_to_close_fds_secs` field will be a floating point value. Otherwise, the value will be null. It would be interesting to look at values close to the leak-timeout limit.

**Comment by sunshowers** [comment]
Based on some profiling, I've also made a couple of changes to leak detection in https://github.com/nextest-rs/nextest/commit/dc6a40ed5eec8abd623887073f532d50ca5e54d9. Curious if it works better for you now, particularly the `biased` change.
